File: rustnlpjava/src/main/java/dev/gigapixel/rustnlp/EncodeResult.java
package dev.gigapixel.rustnlp;

import dev.gigapixel.rustnlp.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class EncodeResult {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class EncodeResultCleaner implements Runnable {

        MemorySegment segment;
        EncodeResultCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            rustnlp_h.EncodeResult_destroy(this.segment);
        }
    }

    EncodeResult() {}
    EncodeResult(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    
    public TokenSpan getByIdx(int i) {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var iNative = i;
            var nativeVal = rustnlp_h.EncodeResult_get_by_idx(arena, internal, iNative);
            
            var returnVal = new TokenSpan(nativeVal);
            return returnVal;
                    
        }
    }
    
    public int len() {
        
        
        var nativeVal = rustnlp_h.EncodeResult_len(internal);
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
    public boolean isEmpty() {
        
        
        var nativeVal = rustnlp_h.EncodeResult_is_empty(internal);
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
    public PairSeq pairSeq() {
        
        
        var nativeVal = rustnlp_h.EncodeResult_pair_seq(internal);
        
        List<Object> selfEdges = List.of();
        
        
        
        var returnVal = new PairSeq(nativeVal, selfEdges);
        return returnVal;
                
    }
    
}

File: rustnlpkotlin/src/main/kotlin/dev/gigapixel/rustnlp/EncodeResult.kt
package dev.gigapixel.rustnlp;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure


internal interface EncodeResultLib: Library {
    fun EncodeResult_destroy(handle: Pointer)
    fun EncodeResult_get(handle: Pointer, i: Int): OptionTokenSpanNative
    fun EncodeResult_get_by_idx(handle: Pointer, i: Int): TokenSpanNative
    fun EncodeResult_len(handle: Pointer): Int
    fun EncodeResult_is_empty(handle: Pointer): Byte
    fun EncodeResult_pair_seq(handle: Pointer): Pointer
}

class EncodeResult internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>
)  {

    internal class EncodeResultCleaner(val handle: Pointer, val lib: EncodeResultLib) : Runnable {
        override fun run() {
            lib.EncodeResult_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<EncodeResultLib> = EncodeResultLib::class.java
        internal val lib: EncodeResultLib = Native.load("rustnlp", libClass)
    }
    
    internal fun getInternal(i: Int): TokenSpan? {
        
        val returnVal = lib.EncodeResult_get(handle, i);
        
        val intermediateOption = returnVal.option() ?: return null
        
        val returnStruct = TokenSpan(intermediateOption)
        return returnStruct
                                
    }
    
    fun getByIdx(i: Int): TokenSpan {
        
        val returnVal = lib.EncodeResult_get_by_idx(handle, i);
        
        val returnStruct = TokenSpan(returnVal)
        return returnStruct
    }
    
    fun len(): Int {
        
        val returnVal = lib.EncodeResult_len(handle);
        return (returnVal)
    }
    
    fun isEmpty(): Boolean {
        
        val returnVal = lib.EncodeResult_is_empty(handle);
        return (returnVal > 0)
    }
    
    fun pairSeq(): PairSeq {
        
        val returnVal = lib.EncodeResult_pair_seq(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = PairSeq(handle, selfEdges)
        CLEANER.register(returnOpaque, PairSeq.PairSeqCleaner(handle, PairSeq.lib));
        return returnOpaque
    }

    operator fun get(index: Int): TokenSpan {
        val returnVal = getInternal(index)
        if (returnVal == null) {
            throw IndexOutOfBoundsException("Index $index is out of bounds.")
        } else {
            return returnVal
        }
    }

}
